File: money.py
# Ngôn ngữ: Python
#
# Công dụng: Money minh họa sự khác nhau giữa "is" (so sánh ĐỊA CHỈ — có phải
# cùng một object trong bộ nhớ) và == (so sánh NỘI DUNG — gọi __eq__ mà ta ghi
# đè), kèm quy ước bắt buộc: ghi đè __eq__ thì phải ghi đè __hash__.
#
# Tại sao cần học: dùng "is" thay == (hoặc ngược lại) là lỗi Python kinh điển —
# bug chập chờn vì đôi khi hai biến tình cờ trỏ cùng object. Ghi đè __eq__ mà
# quên __hash__ làm object không bỏ được vào set/dict (Python tự vô hiệu hóa
# __hash__ khi chỉ ghi đè __eq__).


class Money:  # khai báo class — khuôn mẫu để tạo ra các object Money
    def __init__(self, amount, currency):  # constructor — chạy khi gọi "Money(...)"
        if amount < 0:  # validate: chặn tiền âm
            raise ValueError("Số tiền không được âm")  # ném lỗi ngay tại biên
        if not currency:  # validate: loại tiền bắt buộc
            raise ValueError("Loại tiền không được rỗng")  # ném lỗi ngay tại biên
        self._amount = amount  # số tiền (đơn vị nhỏ nhất) — dùng số nguyên, không dùng float cho tiền
        self._currency = currency  # loại tiền, vd. "VND", "USD"

    def __eq__(self, other):  # ghi đè == thừa hưởng từ object (bản mặc định chỉ so địa chỉ, y hệt "is")
        if not isinstance(other, Money):  # khác class (hoặc None) → không thể bằng nhau
            return NotImplemented  # quy ước Python: trả NotImplemented để nhường quyền so sánh cho phía bên kia
        return self._amount == other._amount and self._currency == other._currency  # bằng nhau khi CÙNG số tiền và CÙNG loại tiền

    def __hash__(self):  # quy ước bắt buộc: đã ghi đè __eq__ thì phải ghi đè __hash__
        return hash((self._amount, self._currency))  # băm tuple gồm đúng các field mà __eq__ so sánh — thư viện chuẩn lo phần còn lại


# ---- Self-check ----
if __name__ == "__main__":  # chỉ chạy khi thực thi trực tiếp "python money.py"
    a = Money(100_000, "VND")  # tạo object thứ nhất
    b = Money(100_000, "VND")  # tạo object thứ hai — CÙNG giá trị nhưng KHÁC object trong bộ nhớ

    assert a is not b, "hai lần khởi tạo phải tạo hai object khác nhau"  # "is" so địa chỉ: hai object riêng biệt
    assert a == b, "cùng số tiền + loại tiền phải bằng nhau"  # == gọi __eq__ ta ghi đè: so nội dung
    assert hash(a) == hash(b), "bằng nhau phải cùng hash"  # quy ước __eq__/__hash__
    assert a != Money(100_000, "USD"), "khác loại tiền không được bằng nhau"  # 100k VND khác 100k USD
    assert a != Money(50_000, "VND"), "khác số tiền không được bằng nhau"  # 100k khác 50k
    assert a != None, "so với None phải trả về False"  # so với None không được ném lỗi (NotImplemented → Python tự kết luận khác nhau)

    print("OK")  # in "OK" nếu mọi assert ở trên đều đúng

File: Money.cpp
/*
 * Ngôn ngữ: C++
 *
 * Công dụng: Money minh họa sự khác nhau giữa so sánh ĐỊA CHỈ (&a == &b — có
 * phải cùng một object trong bộ nhớ) và so sánh NỘI DUNG (a == b qua nạp chồng
 * operator== — hai object có cùng giá trị nghiệp vụ không), kèm quy ước đi cùng:
 * muốn bỏ Money vào unordered_set/unordered_map thì phải cung cấp hàm băm
 * (std::hash) nhất quán với operator==.
 *
 * Tại sao cần học: khác Java (== có sẵn nhưng so địa chỉ), C++ KHÔNG tự sinh
 * operator== cho class — quên nạp chồng là code không biên dịch; còn nạp chồng
 * lệch với hàm băm làm object "biến mất" trong unordered_set/unordered_map.
 */
#include <cstdint>    // thư viện chuẩn: std::int64_t — số nguyên 64 bit cho tiền
#include <cstdlib>    // thư viện chuẩn: std::exit để dừng chương trình khi self-check thất bại
#include <functional> // thư viện chuẩn: std::hash để sinh mã băm
#include <iostream>   // thư viện chuẩn: std::cout/std::cerr để in ra màn hình
#include <stdexcept>  // thư viện chuẩn: std::invalid_argument để báo lỗi dữ liệu sai
#include <string>     // thư viện chuẩn: std::string cho loại tiền

class Money { // khai báo class — khuôn mẫu để tạo ra các object Money
private:
    const std::int64_t amount;  // số tiền (đơn vị nhỏ nhất) — dùng số nguyên, không dùng double cho tiền
    const std::string currency; // loại tiền, vd. "VND", "USD"

public:
    Money(std::int64_t amount, const std::string& currency) : amount(amount), currency(currency) { // constructor — chạy khi tạo object
        if (amount < 0) throw std::invalid_argument("Số tiền không được âm"); // validate: chặn tiền âm
        if (currency.empty()) throw std::invalid_argument("Loại tiền không được rỗng"); // validate: loại tiền bắt buộc
    }

    bool operator==(const Money& other) const { // nạp chồng == : so sánh NỘI DUNG — C++ không tự sinh, phải tự viết
        return amount == other.amount && currency == other.currency; // bằng nhau khi CÙNG số tiền và CÙNG loại tiền
    }

    bool operator!=(const Money& other) const { // nạp chồng != cho đủ cặp — luôn định nghĩa qua == để hai toán tử không lệch nhau
        return !(*this == other); // phủ định của == (*this = object hiện tại)
    }

    std::size_t hashValue() const { // mã băm — hai object == nhau PHẢI trả về cùng mã băm (như quy ước equals/hashCode của Java)
        return std::hash<std::int64_t>{}(amount) ^ (std::hash<std::string>{}(currency) << 1); // kết hợp mã băm của đúng các field mà operator== so sánh
    }
};

// ---- Self-check ----
static void check(bool ok, const std::string& msg) { // hàm phụ trợ: nếu "ok" sai thì báo lỗi kèm thông điệp "msg"
    if (!ok) { std::cerr << "FAIL: " << msg << "\n"; std::exit(1); } // in chỗ sai rồi dừng ngay
}

int main() { // điểm bắt đầu chạy chương trình sau khi biên dịch
    Money a(100'000, "VND"); // tạo object thứ nhất
    Money b(100'000, "VND"); // tạo object thứ hai — CÙNG giá trị nhưng KHÁC object trong bộ nhớ

    check(&a != &b, "hai lần khai báo phải tạo hai object khác nhau");        // so địa chỉ (&) : hai object riêng biệt trong bộ nhớ
    check(a == b, "cùng số tiền + loại tiền phải bằng nhau");                  // operator== so nội dung: cùng giá trị nghiệp vụ
    check(a.hashValue() == b.hashValue(), "bằng nhau phải cùng mã băm");       // quy ước ==/hash: bằng nhau thì băm giống nhau
    check(a != Money(100'000, "USD"), "khác loại tiền không được bằng nhau");  // 100k VND khác 100k USD
    check(a != Money(50'000, "VND"), "khác số tiền không được bằng nhau");     // 100k khác 50k

    std::cout << "OK\n"; // in "OK" nếu mọi check ở trên đều đúng
    return 0; // báo chương trình kết thúc thành công
}
